### PR TITLE
loaddata: split functionalities to prepare concurrent load

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref #40499 

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: executor/load_data.go [hunk 1]
```diff
@@ -250,126 +246,152 @@ type loadRemoteInfo struct {
...
+	// I have no confident that all needed variables are copied :(
+	fromVars := userSctx.GetSessionVars()
+	toVars := sysSession.GetSessionVars()
+	toVars.User = fromVars.User
+	toVars.CurrentDB = fromVars.CurrentDB
+	toVars.SQLMode = fromVars.SQLMode
+	if !controller.Restrictive {
+		setNonRestrictiveFlags(toVars.StmtCtx)
+	}
+	sysSession.GetSessionVars().StmtCtx.InitSQLDigest(
```
Comment (lance6716) [inline executor/load_data.go:364]: @mornyx PTAL

**Comment by lance6716** [comment]
unit test passed https://do.pingcap.net/jenkins/blue/organizations/jenkins/pingcap%2Ftidb%2Fghpr_unit_test/detail/ghpr_unit_test/20711/pipeline

will remove debug message

File: executor/load_data.go [hunk 2]
```diff
@@ -89,23 +91,11 @@ func (e *LoadDataExec) Next(ctx context.Context, req *chunk.Chunk) (err error) {
...
 	switch e.FileLocRef {
 	case ast.FileLocServerOrRemote:
-		u, err2 := storage.ParseRawURL(e.loadDataWorker.GetInfilePath())
-		if err2 != nil {
-			return exeerrors.ErrLoadDataInvalidURI.GenWithStackByArgs(err2.Error())
-		}
-		path := strings.Trim(u.Path, "/")
-		u.Path = ""
-		b, err2 := storage.ParseBackendFromURL(u, nil)
+		b, path, err2 := e.loadDataWorker.resolveRemoteStorage()
```
Comment (D3Hunter) [inline executor/load_data.go:94]: change it too in my pr😅

File: executor/asyncloaddata/progress.go [hunk 3]
```diff
@@ -24,14 +28,25 @@ type Progress struct {
 	// SimpleSeekerOnReadCloser on MySQL client connection which doesn't support
 	// it.
 	SourceFileSize int64
-	// LoadedFileSize is the size of the data that has been loaded in bytes.
+	// LoadedFileSizeSetter is used by multiple workers to record the progress,
+	// its value is loaded into LoadedFileSize when String().
+	LoadedFileSizeSetter atomic.Int64 `json:"-"`
```
Comment (D3Hunter) [inline executor/asyncloaddata/progress.go:33]: can merge with below one?

File: executor/load_data.go [hunk 4]
```diff
@@ -727,285 +853,164 @@ func (e *LoadDataWorker) commitWork(ctx context.Context) (err error) {
...
-			if fieldMappings[i].Column == nil {
-				setVar(fieldMappings[i].UserVar.Name, nil)
-				continue
-			}
+// mergeAndSetMessage merges stats from all used session context and sets info
+// message(ERR_LOAD_INFO) generated by LOAD statement to UserSctx, it is
+// public for test.
+func (e *LoadDataWorker) mergeAndSetMessage() string {
+	encodeStmtCtx := e.encodeWorker.ctx.GetSessionVars().StmtCtx
+	numAffected := encodeStmtCtx.AffectedRows()
```
Comment (D3Hunter) [inline executor/load_data.go:978]: is this always 0?

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: executor/asyncloaddata/progress.go (hunk 3 above)
Comment (lance6716) [reply executor/asyncloaddata/progress.go:33]: Do you mean we merge `LoadedFileSizeSetter` with `LoadedFileSize`? I think I can write a customized json Marshaller/Unmarshaller to do it

File: executor/load_data.go (hunk 4 above)
Comment (lance6716) [reply executor/load_data.go:978]: Seems yes, I'll check CI without counting encodeStmtCtx.

File: executor/load_data.go [hunk 5]
```diff
@@ -689,33 +690,158 @@ func (e *LoadDataWorker) processStream(
...
+			}
+			return exeerrors.ErrLoadDataCantRead.GenWithStackByArgs(
+				err.Error(),
+				"Only the following formats delimited text file (csv, tsv), parquet, sql are supported. Please provide the valid source file(s)",
+			)
+		}
+		// rowCount will be used in fillRow(), last insert ID will be assigned according to the rowCount = 1.
+		// So should add first here.
+		w.rowCount++
+		r, err := w.parserData2TableData(ctx, parser.LastRow().Row)
```
Comment (GMHDBJD) [inline executor/load_data.go:736]: do we need to call parser.RecycleRow?

File: executor/asyncloaddata/progress.go [hunk 6]
```diff
@@ -24,21 +28,69 @@ type Progress struct {
 	// SimpleSeekerOnReadCloser on MySQL client connection which doesn't support
 	// it.
 	SourceFileSize int64
-	// LoadedFileSize is the size of the data that has been loaded in bytes.
-	LoadedFileSize int64
+	// LoadedFileSize is the size of the data that will be loaded in bytes. It's
+	// larger than the actual loaded data size, but due to the fact that reading
+	// is once-a-block and a block may generate multiple tasks that are
+	// concurrently executed, we can't know the actual loaded data size easily.
+	LoadedFileSize atomicInt64JSONWrapper
```
Comment (D3Hunter) [inline executor/asyncloaddata/progress.go:35]: https://github.com/uber-go/atomic/pull/68 can we use this

**Review by D3Hunter — APPROVED** [review]

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Pull request merged**